Cache/IO.lean

```diff
@@ -136,6 +156,45 @@ def validateCurl : IO Bool := do
     | _ => throw $ IO.userError "Invalidly formatted version of `curl`"
   | _ => throw $ IO.userError "Invalidly formatted response from `curl --version`"
 
+def Version := Nat × Nat × Nat
+  deriving Inhabited, DecidableEq
+
+instance : Ord Version := let _ := @lexOrd; lexOrd
+instance : LE Version := leOfOrd
+
+def parseVersion (s : String) : Option Version := do
+  let [maj, min, patch] := s.splitOn "." | none
+  some (maj.toNat!, min.toNat!, patch.toNat!)
+
+def validateLeanTar : IO Unit := do
+  if (← LEANTARBIN.pathExists) then return
+  if let some version ← some <$> runCmd "leantar" #["--version"] <|> pure none then
+    let "leantar" :: v :: _ := version.splitOn " "
+      | throw $ IO.userError "Invalidly formatted response from `leantar --version`"
+    let some v := parseVersion v | throw $ IO.userError "Invalidly formatted version of `leantar`"
+    -- currently we need exactly one version of leantar, change this to reflect compatibility
+    if v = (parseVersion LEANTARVERSION).get! then return
+  let win := System.Platform.getIsWindows ()
+  let target ← if win then
+    pure "x86_64-pc-windows-msvc"
+  else
+    let mut arch ← (·.trim) <$> runCmd "uname" #["-m"] false
+    if arch = "arm64" then arch := "aarch64"
+    unless arch ∈ ["x86_64", "aarch64"] do
+      throw $ IO.userError s!"unsupported architecture {arch}"
+    pure <|
+      if System.Platform.getIsOSX () then s!"{arch}-apple-darwin"
+      else s!"{arch}-unknown-linux-musl"
+  IO.println s!"leantar is too old; downloading more recent version"
+  IO.FS.createDirAll IO.CACHEDIR
+  let ext := if win then "zip" else "tar.gz"
+  let _ ← runCmd "curl" #[
+    s!"https://github.com/digama0/leangz/releases/download/v{LEANTARVERSION}/leantar-v{LEANTARVERSION}-{target}.{ext}",
+    "-L", "-o", s!"{LEANTARBIN}.{ext}"]
+  let _ ← runCmd "tar" #["-xf", s!"{LEANTARBIN}.{ext}",
+    "-C", IO.CACHEDIR.toString, "--strip-components=1"]
```

Comment:
Does this work on windows (i.e., with zip)?

Reply:
It seems to work... not sure how universally available `tar` is on windows but this exact command was tested and works on at least a few machines.

Reply:
:scream: 

> The old cache was also using tar for unpacking on windows and I don't think we have received any complaints about it.

The old cache used tar to unpack tar files.  This line here uses tar to unpack zip files.  But if it works. :shrug: